### PR TITLE
[NFCI][SYCL] Refactor sycl/test-e2e/Config/select_device.cpp

### PR DESCRIPTION
**Pull request by aelovikov-intel**

### CODE REVIEW

File: sycl/test-e2e/Config/select_device.cpp [hunk 1]
```diff
@@ -167,285 +167,215 @@ static std::vector<DevDescT> getAllowListDesc(std::string allowList) {
   return decDescs;
 }
 
-int main() {
-  bool passed = false;
-  std::string sycl_be("opencl");
-  if (getenv("ONEAPI_DEVICE_SELECTOR")) {
-    sycl_be = getenv("ONEAPI_DEVICE_SELECTOR");
+bool is_known_be(backend be) {
+    backend known_bes[] = {backend::opencl, backend::ext_oneapi_level_zero,
```
Comment (cperkinsintel) [inline sycl/test-e2e/Config/select_device.cpp:171]: do we have to worry about esmid emulator backend?

Comment (aelovikov-intel) [reply sycl/test-e2e/Config/select_device.cpp:171]: No entry - no problems :D

That part is NFC and I tried to follow what the tests did before my change (as long as it wasn't an obvious bug/mistake).

File: sycl/test-e2e/Config/select_device.cpp [hunk 2]
```diff
@@ -456,510 +386,274 @@ int main() {
...
-              passed = true;
-              break;
+          if (!dev.has(aspect::gpu))
+            continue;
+          std::string name = dev.get_info<info::device::name>();
+          addEscapeSymbolToSpecialCharacters(name);
+          std::string ver = dev.get_info<info::device::driver_version>();
+          if (is_known_be(plt.get_backend())) {
+            if (count > 0) {
+              std::cout << "|";
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Config/select_device.cpp:501]: This is a bugfix - LHS has `" | "` at line 653.

File: sycl/test-e2e/Config/select_device.cpp [hunk 3]
```diff
@@ -456,510 +386,274 @@ int main() {
...
+        // intel/llvm-test-suite with removal of 1st parameter of the vector,
+        // and transformation of std::vector<std::string> to std::string
+        const std::vector<std::string> expectedMsgs{
+            "Malformed syntax in SYCL_DEVICE_ALLOWLIST",
+            "Key PlatformVersion of SYCL_DEVICE_ALLOWLIST should have value "
+            "which starts with {{"};
+        const std::string gotMessage(E.what());
+        std::cout << "Got message: " << gotMessage << std::endl;
+        // FIXME:
+        return 0;
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Config/select_device.cpp:647]: Whoops....

File: sycl/test-e2e/Config/select_device.cpp [hunk 4]
```diff
@@ -456,510 +386,274 @@ int main() {
...
+    } else {
+      try {
+        sycl::queue deviceQueue(gpu_selector_v);
+        device dev = deviceQueue.get_device();
+        const auto &plt = dev.get_platform();
+      } catch (sycl::exception &E) {
+        std::cout << "Caught exception: " << E.what() << std::endl;
+        if (E.what() ==
+            "Key DriverVersion of SYCL_DEVICE_ALLOWLIST should have value which starts with {{ -30 (PI_ERROR_INVALID_VALUE)"sv)
+          return 0;
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Config/select_device.cpp:611]: This (and similar changes in few other places) addresses TODO/workaround from the LHS.

File: sycl/test-e2e/Config/select_device.cpp [hunk 5]
```diff
@@ -456,510 +386,274 @@ int main() {
...
+            if ((plt.get_backend() == backend::opencl) ||
+                (plt.get_backend() == backend::ext_oneapi_level_zero)) {
               if ((pos = ver.find(".")) == std::string::npos) {
                 throw std::runtime_error("Malformed syntax in version string");
               }
               pos++;
               size_t start = pos;
+              // FIXME: What is going on here?!! It's insanity to expect a
+              // different result than few lines above. Is that a weirdest
+              // possible way to write "pos -= 1;"???
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Config/select_device.cpp:398]: This comment is new.

File: sycl/test-e2e/Config/select_device.cpp [hunk 6]
```diff
@@ -167,285 +167,215 @@ static std::vector<DevDescT> getAllowListDesc(std::string allowList) {
...
+      const auto &plt = dev.get_platform();
+      std::cout << "Platform: " << plt.get_info<info::platform::name>()
+                << std::endl;
+      std::cout << "Platform Version: "
+                << plt.get_info<info::platform::version>() << std::endl;
+      for (const DevDescT &desc : components) {
+        std::cout << "Desc pltName: " << desc.platName << std::endl;
+        std::cout << "Desc pltVer: " << desc.platVer << std::endl;
+        if ((std::regex_match(plt.get_info<info::platform::name>(),
+                              std::regex(desc.platName))) &&
```
Comment (cperkinsintel) [inline sycl/test-e2e/Config/select_device.cpp:295]: if the platform name contains any regex reserved character, like `.  ?  * [ ] ^ $ % \ -`  this won't work.   I haven't use the std:: regex library much, but in other languages there is "RegexEscape" or equivalent that lets you convert any string into an arbitrary "safe" regex that just matches the intended string.

Comment (aelovikov-intel) [reply sycl/test-e2e/Config/select_device.cpp:295]: I'm not modifying this either.

File: sycl/test-e2e/Config/select_device.cpp (hunk 5 above)
Comment (cperkinsintel) [reply sycl/test-e2e/Config/select_device.cpp:398]: It's my second  favorite comment in the code base, my most being  " // Drunk, fix later. "

File: sycl/test-e2e/Config/select_device.cpp (hunk 6 above)
Comment (cperkinsintel) [reply sycl/test-e2e/Config/select_device.cpp:295]: do we even need the regex then? Wouldn't just a standard string search  or position work?

**Review by cperkinsintel — APPROVED** [review]
LGTM - this test has always been a bit strange. It's still strange, but perhaps less so.

**Pull request merged**